openssl/openssl.go

```diff
@@ -129,18 +131,47 @@ func SetFIPS(enabled bool) error {
 		if !providerAvailable(props) {
 			// If not, fallback to provName provider.
 			if C.go_openssl_OSSL_PROVIDER_load(nil, provName) == nil {
-				return fail("OSSL_PROVIDER_try_load")
+				return newOpenSSLError("OSSL_PROVIDER_try_load")
 			}
 			// Make sure we now have a provider available.
 			if !providerAvailable(props) {
 				return fail("SetFIPS(" + strconv.FormatBool(enabled) + ") not supported")
 			}
 		}
 		if C.go_openssl_EVP_set_default_properties(nil, props) != 1 {
-			return fail("EVP_set_default_properties")
+			return newOpenSSLError("EVP_set_default_properties")
 		}
 		return nil
 	default:
 		panic(errUnsupportedVersion())
 	}
 }
+
+func newOpenSSLError(msg string) error {
+	var b strings.Builder
+	b.WriteString(msg)
+	b.WriteString("\nopenssl error(s):")
+	for {
+		var (
+			e    C.ulong
+			file *C.char
+			line C.int
+		)
+		switch vMajor {
+		case 1:
+			e = C.go_openssl_ERR_get_error_line(&file, &line)
+		case 3:
+			e = C.go_openssl_ERR_get_error_all(&file, &line, nil, nil, nil)
+		default:
+			panic(errUnsupportedVersion())
+		}
+		if e == 0 {
+			break
+		}
+		b.WriteByte('\n')
+		var buf [256]byte
+		C.go_openssl_ERR_error_string_n(e, (*C.char)(unsafe.Pointer(&buf[0])), C.size_t(len(buf)))
+		b.WriteString(string(buf[:]) + "\n\t" + C.GoString(file) + ":" + strconv.Itoa(int(line)))
```

Comment:
Since we're already using a builder, would it be more efficient to add these pieces one-by-one with `b.WriteString`?

Reply:
It would be more efficient, but also uglier:

```go
b.WriteString(string(buf[:]) + "\n\t" + C.GoString(file) + ":" + strconv.Itoa(int(line)))

// vs

b.WriteByte('\n')
b.Write(buf[:])
b.WriteString("\n\t")
b.WriteString(C.GoString(file))
b.WriteByte(':')
b.WriteString(strconv.Itoa(int(line)))
```

I would rather keep it one-liner 😄